tss-esapi/src/abstraction/public.rs

```diff
@@ -0,0 +1,116 @@
+use crate::interface_types::ecc::EccCurve;
+use crate::structures::{Public, RsaExponent};
+use crate::{Error, WrapperErrorKind};
+
+use core::convert::TryFrom;
+use oid::ObjectIdentifier;
+use picky_asn1::bit_string::BitString;
+use picky_asn1::wrapper::{IntegerAsn1, OctetStringAsn1};
+use picky_asn1_x509::{
+    AlgorithmIdentifier, EcParameters, EcPoint, PublicKey, RsaPublicKey, SubjectPublicKeyInfo,
+};
+use serde::{Deserialize, Serialize};
+
+#[derive(Serialize, Deserialize, Debug, PartialEq, Clone)]
+pub enum DecodedKey {
+    RsaPublicKey(RsaPublicKey),
+    EcPoint(EcPoint),
+}
+
+impl TryFrom<Public> for DecodedKey {
+    type Error = Error;
+
+    fn try_from(value: Public) -> Result<Self, Self::Error> {
+        public_to_decoded_key(&value)
+    }
+}
+
+impl TryFrom<Public> for SubjectPublicKeyInfo {
+    type Error = Error;
+
+    /// Converts [`crate::structures::Public::Rsa`] and [`crate::structures::Public::Ecc`] to [`picky_asn1_x509::SubjectPublicKeyInfo`].
+    ///
+    /// # Details
+    /// The result can be used to convert TPM public keys to DER using `picky_asn1_der`.
+    ///
+    /// # Errors
+    /// * if other instances of [`crate::structures::Public`] are used `UnsupportedParam` will be returned.
+    fn try_from(value: Public) -> Result<Self, Self::Error> {
+        let decoded_key = public_to_decoded_key(&value)?;
+
+        match (value, decoded_key) {
+            (Public::Rsa { .. }, DecodedKey::RsaPublicKey(key)) => Ok(SubjectPublicKeyInfo {
+                algorithm: AlgorithmIdentifier::new_rsa_encryption(),
```

Comment:
@ionut-arm do you know if this is hash algorithm independent or do we actually need to use the specific identifiers for SHA1, SHA256 etc.?

Reply:
As far as I know/understand, this identifier is supposed to match up to the type of the key that follows in `SubjectPublicKeyInfo`. If you look at [the RFC](https://datatracker.ietf.org/doc/html/rfc8017#appendix-A.1), it does mention [an OID](http://oid-info.com/get/1.2.840.113549.1.1.1) (`pkcs-1 1`) that applies to the `RSAPublicKey` format. This is [the same one](https://github.com/Devolutions/picky-rs/blob/master/picky-asn1-x509/src/oids.rs#L54) that you get here from `picky-asn1-x509` as well, so it should be good. ([this](https://jpassing.com/2021/11/30/using-pem-to-encode-public-keys/) blog post also finds the same.) I think this is correct!